### PR TITLE
PR: Add support for Jupyter-client >= 7

### LINKED ISSUE #16762: _kill_kernel changed in jupyter_clients 7.0.6
### What steps will reproduce the problem?

close spyder

### Traceback
```python-traceback
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/manager.py", line 380, in _async_finish_shutdown
    await asyncio.wait_for(
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 498, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/utils.py", line 26, in wrapped
    raise e
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/utils.py", line 23, in wrapped
    return loop.run_until_complete(future)
  File "/usr/local/lib/python3.8/site-packages/nest_asyncio.py", line 70, in run_until_complete
    return f.result()
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 280, in __step
    result = coro.send(None)
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/manager.py", line 443, in _async_shutdown_kernel
    await ensure_async(self.finish_shutdown(restart=restart))
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/utils.py", line 26, in wrapped
    raise e
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/utils.py", line 23, in wrapped
    return loop.run_until_complete(future)
  File "/usr/local/lib/python3.8/site-packages/nest_asyncio.py", line 70, in run_until_complete
    return f.result()
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 280, in __step
    result = coro.send(None)
  File "/usr/local/lib/python3.8/site-packages/jupyter_client/manager.py", line 386, in _async_finish_shutdown
    await ensure_async(self._kill_kernel(restart=restart))
TypeError: _kill_kernel() got an unexpected keyword argument 'restart'
```

## Versions

* Spyder version: 6.0.0.dev0 a84c40648
* Python version: 3.8.6
* Qt version: 5.12.10
* PyQt5 version: 5.12.3
* Operating System: Darwin 20.6.0

### Dependencies

```

# Mandatory:
applaunchservices >=0.1.7              :  0.2.1 (OK)
atomicwrites >=1.2.0                   :  1.4.0 (OK)
chardet >=2.0.0                        :  3.0.4 (OK)
cloudpickle >=0.5.0                    :  2.0.0 (OK)
cookiecutter >=1.6.0                   :  1.7.2 (OK)
diff_match_patch >=20181111            :  20200713 (OK)
intervaltree >=3.0.2                   :  3.1.0 (OK)
IPython >=7.6.0                        :  7.29.0 (OK)
jedi >=0.17.2;<0.19.0                  :  0.18.0 (OK)
jellyfish >=0.7                        :  None (OK)
jsonschema >=3.2.0                     :  3.2.0 (OK)
keyring >=17.0.0                       :  21.4.0 (OK)
nbconvert >=4.0                        :  6.0.7 (OK)
numpydoc >=0.6.0                       :  1.1.0 (OK)
parso >=0.7.0;<0.9.0                   :  0.8.2 (OK)
pexpect >=4.4.0                        :  4.8.0 (OK)
pickleshare >=0.4                      :  0.7.5 (OK)
psutil >=5.3                           :  5.7.2 (OK)
pygments >=2.0                         :  2.10.0 (OK)
pylint >=2.5.0                         :  2.6.0 (OK)
pyls_spyder >=0.4.0                    :  0.4.0 (OK)
pylsp >=1.2.2;<1.3.0                   :  1.1.0 (NOK)
pylsp_black >=1.0.0                    :  None (OK)
qdarkstyle =3.0.2                      :  3.0.2 (OK)
qstylizer >=0.1.10                     :  0.1.10 (OK)
qtawesome >=1.0.2                      :  1.0.2 (OK)
qtconsole >=5.1.0                      :  5.2.0.dev0 (OK)
qtpy >=1.5.0                           :  1.9.0 (OK)
rtree >=0.9.7                          :  None (OK)
setuptools >=49.6.0                    :  50.3.0 (OK)
sphinx >=0.6.6                         :  3.2.1 (OK)
spyder_kernels >=2.1.1;<2.2.0          :  2.2.0.dev0 (OK)
textdistance >=4.2.0                   :  4.2.0 (OK)
three_merge >=0.1.1                    :  0.1.1 (OK)
watchdog >=0.10.3                      :  0.10.3 (OK)
zmq >=17                               :  22.3.0 (OK)

# Optional:
cython >=0.21                          :  0.29.21 (OK)
matplotlib >=2.0.0                     :  3.4.2 (OK)
numpy >=1.7                            :  1.21.0 (OK)
pandas >=1.1.1                         :  1.1.3 (OK)
scipy >=0.17.0                         :  1.5.2 (OK)
sympy >=0.7.3                          :  1.6.2 (OK)

# Spyder plugins:
spyder_boilerplate.spyder.plugin 0.0.1 :  0.0.1 (OK)
```

### PR DESCRIPTION
**Pull request by bnavigator**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Follow-up on https://github.com/spyder-ide/spyder-kernels/pull/328
Fixes #16762.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @bnavigator

### CODE REVIEW

**Comment by bnavigator** [comment]
I think this is ready.

I reverted the "test commits", which force jupyter_client 7 and clone the subrepo from https://github.com/spyder-ide/spyder-kernels/pull/328. The checks on 46556ad show that it works. (The Mac App Bundle error looks unrelated)

**Comment by ccordoba12** [comment]
> I think this is ready.

Thanks @bnavigator for your help!

> I reverted the "test commits", which force jupyter_client 7 and clone the subrepo from spyder-ide/spyder-kernels#328

I think you should remove them and leave commit 46556ad as the last one on this PR, so that we continue our testing with Jupyter-client 7.

> (The Mac App Bundle error looks unrelated)

Please merge with the latest 5.x or rebase on top of it to get the fixes for that.

**Comment by ccordoba12** [comment]
By the way, I'm a bit worried about issue https://github.com/jupyter/jupyter_client/issues/715 because on Windows kernels take a llong time to start (I've seen them up to a minute to do it).

So, we'll have to wait until it's fixed (there's a PR for that already). After that is part of an official release, I think we can merge this.

**Comment by bnavigator** [comment]
> I think you should remove them and leave commit [46556ad](https://github.com/spyder-ide/spyder/commit/46556ad2431dbc2d966fde7d9c2788532044eca4) as the last one on this PR, so that we continue our testing with Jupyter-client 7.

I am not sure if I understand you correctly. Is eb9c9b8 the way you intended it?

**Comment by bnavigator** [comment]
Sure. But it means you definitely need spyder-kernels updated. And you probably want to use your own branch instead of the PR eventually.

**Comment by ccordoba12** [comment]
I understand your concern now. That's why:

1. We have a subrepo for spyder-kernels here, which is a copy of its repo that we sync with git for our respective branches here (`2.x` with `5.x` and `master` with `master`). Our tests run with that subrepo, not with the installed version.
2. We release a new version of spyder-kernels a day or two before releasing a new one of Spyder.

File: external-deps/spyder-kernels/.gitrepo [hunk 1]
```diff
@@ -5,8 +5,8 @@
 ;
 [subrepo]
 	remote = https://github.com/spyder-ide/spyder-kernels.git
-	branch = 2.x
-	commit = ac40350c9260e2e5d73f0fc3f8a9cc15da8d8a03
-	parent = 01421c2ae26c7a3861fc4c912f295f9f2b2c1c1f
+	branch = client7
```
Comment (bnavigator) [inline external-deps/spyder-kernels/.gitrepo:8]: I mean, you probably don't want to merge while this is referenced.

Comment (ccordoba12) [reply external-deps/spyder-kernels/.gitrepo:8]: Right, I'll merge your corresponding PR in spyder-kernels, then ask you to resync our subrepo here, then merge.

The process is explained in our [contributing guide](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md).

**Comment by ccordoba12** [comment]
@bnavigator, I updated your PR in spyder-kernels and this one in order to merge them for Spyder 5.2.0 (to be released on Wednesday). I was not planning to do that, but our tests in spyder-kernels are constantly failing with jupyter_client 6.1.12, so this is a necessity now.

Sorry for the rush and hope it's ok for you.

**Review by ccordoba12 — APPROVED** [review]
Thanks @bnavigator! I really appreciate your help with this.

**Pull request merged**